java/arcs/core/host/ArcState.kt

```diff
@@ -99,6 +104,11 @@ enum class ParticleState {
     val failed: Boolean
         get() = failedStates.contains(this)
 
+    /**
+     * When a [Particle] moves to a failed state, the cause may be attached here.
+     */
+    var errorCause: Exception? = null
```

Comment:
Enum's are static singletons normally, so assigning to this field I believe overwrites any other enum refs being held in other threads. If you're going to do this, I think you need to convert this enum into a sealed class.

Reply:
Additionally, other issues:

 Exception isn't Parcelizable, so by default, if this is sent via IPC in an Intent or Bundle, either it will cause an error, or this field will end up default initialized to null.  Look in ArcHostHelper.runWithResult method for how Enum's are sent across the wire. Enums are serialized as ordinal only.  There is an existing method for propagating exceptions, EXTRA_OPERATION_EXCEPTION and EXTRA_OPERATION_EXCEPTION_STACKTRACE. 

Because Arcs is multiple language and multiprocess (e.g. JsHost in future), these remote exceptions are represented in serialized String format. 

ResultReceiverContinuation in IntentHostAdapter reifies these back into an ArcHostException currently. To make Android work with your pattern (and future JsHosts), I think you'd need to make the ArcState.errorCause a data class that describes the exception, and change the serialization/deserialization logic in runWithResult/ResultReceiverContinuation to marshall/unmarshall this through the Bundle back into the ArcState. 

Look at IntentArcHostAdapter for how it marshalls ArcState for ArcStateChangeHandler callbacks via ResultReceiverStateChangeHandler.

Reply:
Thanks for the info Ray, will update.